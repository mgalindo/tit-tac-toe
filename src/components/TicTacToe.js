import * as React from "react";
import { Alert, Button, View, StyleSheet } from "react-native";
import Board from "./Board";

/*
Assumes the board tiles are numbered like this
0 1 2
3 4 5
6 7 8
*/
const winCombinations = [
  [0, 1, 2],
  [3, 4, 5],
  [6, 7, 8],
  [0, 3, 6],
  [1, 4, 7],
  [2, 5, 8],
  [0, 4, 8],
  [2, 4, 6]
];

export default class TicTacToe extends React.Component {
  getPlayer = moveIndex => {
    const players = ["X", "O"];
    return players[moveIndex % 2];
  };

  getEmptyState = () => {
    return {
      values: new Array(9).fill(" "),
      curMove: 0,
      curPlayer: this.getPlayer(0)
    };
  };

  state = this.getEmptyState();

  moveHandler = (index, value) => {
    const newValues = [...this.state.values];
    newValues[index - 1] = value;
    const curMove = this.state.curMove + 1;
    const curPlayer = this.getPlayer(curMove);
    this.setState(
      { values: newValues, curMove, curPlayer },
      this.processNewState
    );
  };

  getWinner = values => {
    for (let winComb of winCombinations) {
      const [index1, index2, index3] = winComb;
      if (
        values[index1].trim() !== "" &&
        values[index1] === values[index2] &&
        values[index1] === values[index3]
      ) {
        return values[index1];
      }
    }
    return null;
  };

  processNewState = () => {
    const { curMove, values } = this.state;
    const winner = this.getWinner(values);
    if (winner) {
      Alert.alert("The winner is: " + winner, "", [
        { text: "OK", onPress: this.restartGame }
      ]);
    }

    if (!winner && curMove === 9) {
      Alert.alert("Game is a draw", "", [
        { text: "OK", onPress: this.restartGame }
      ]);
    }
  };

  restartGame = () => {
    this.setState(this.getEmptyState());
  };

  render() {
    const { curPlayer, values } = this.state;
    return (
      <View style={styles.container}>
        <Board
          curPlayer={curPlayer}
          values={values}
          onPlayerMove={this.moveHandler}
        />
        <View style={styles.footer}>
          <Button
            style={styles.button}
            color="grey"
            title="Restart Game"
            onPress={this.restartGame}
          />
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1
  },
  footer: {
    flex: 0.2,
    alignItems: "center",
    justifyContent: "center"
  },
  button: {
    backgroundColor: "black"
  }
});
