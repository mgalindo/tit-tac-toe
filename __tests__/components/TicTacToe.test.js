import React from "react";
import { View } from "react-native";
import { render, fireEvent } from "react-native-testing-library";
import TicTacToe from "../../src/components/TicTacToe";

describe("TicTacToe component", () => {
  // let props;
  // let onMockPlayerMove;

  // beforeEach(() => {
  //   onMockPlayerMove = jest.fn();
  //   jest.clearAllMocks();
  //   props = {
  //     values: new Array(9).fill(" "),
  //     curPlayer: "X",
  //     onPlayerMove: onMockPlayerMove
  //   };
  // });

  test("should render empty TicTacToe", () => {
    //expect.assertions(1);
    const { toJSON } = render(<TicTacToe />);
    expect(toJSON()).toMatchSnapshot();
  });

  // test("should render empty TicTacToe", () => {
  //   //expect.assertions(1);
  //   const { getByTestId } = render(<View><TicTacToe testID="game-id"/></View>);

  //   const game = getByTestId("game-id");
  //   const winner game.getWinner();
  // });

});