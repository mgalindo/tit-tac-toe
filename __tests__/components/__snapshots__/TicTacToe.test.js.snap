// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicTacToe component should render empty TicTacToe 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 0.8,
        "flexDirection": "column",
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#4A8305",
              "height": 150,
              "justifyContent": "center",
              "marginBottom": 5,
              "marginLeft": 6,
              "width": 150,
            },
            Object {},
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "opacity": 1,
            }
          }
          testID="box-button"
        >
          <Text
            style={
              Object {
                "fontSize": 40,
                "fontWeight": "bold",
              }
            }
            testID="box-text"
          >
             
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#4A8305",
              "height": 150,
              "justifyContent": "center",
              "marginBottom": 5,
              "marginLeft": 6,
              "width": 150,
            },
            Object {},
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "opacity": 1,
            }
          }
          testID="box-button"
        >
          <Text
            style={
              Object {
                "fontSize": 40,
                "fontWeight": "bold",
              }
            }
            testID="box-text"
          >
             
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#4A8305",
              "height": 150,
              "justifyContent": "center",
              "marginBottom": 5,
              "marginLeft": 6,
              "width": 150,
            },
            Object {},
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "opacity": 1,
            }
          }
          testID="box-button"
        >
          <Text
            style={
              Object {
                "fontSize": 40,
                "fontWeight": "bold",
              }
            }
            testID="box-text"
          >
             
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#4A8305",
              "height": 150,
              "justifyContent": "center",
              "marginBottom": 5,
              "marginLeft": 6,
              "width": 150,
            },
            Object {},
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "opacity": 1,
            }
          }
          testID="box-button"
        >
          <Text
            style={
              Object {
                "fontSize": 40,
                "fontWeight": "bold",
              }
            }
            testID="box-text"
          >
             
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#4A8305",
              "height": 150,
              "justifyContent": "center",
              "marginBottom": 5,
              "marginLeft": 6,
              "width": 150,
            },
            Object {},
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "opacity": 1,
            }
          }
          testID="box-button"
        >
          <Text
            style={
              Object {
                "fontSize": 40,
                "fontWeight": "bold",
              }
            }
            testID="box-text"
          >
             
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#4A8305",
              "height": 150,
              "justifyContent": "center",
              "marginBottom": 5,
              "marginLeft": 6,
              "width": 150,
            },
            Object {},
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "opacity": 1,
            }
          }
          testID="box-button"
        >
          <Text
            style={
              Object {
                "fontSize": 40,
                "fontWeight": "bold",
              }
            }
            testID="box-text"
          >
             
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#4A8305",
              "height": 150,
              "justifyContent": "center",
              "marginBottom": 5,
              "marginLeft": 6,
              "width": 150,
            },
            Object {},
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "opacity": 1,
            }
          }
          testID="box-button"
        >
          <Text
            style={
              Object {
                "fontSize": 40,
                "fontWeight": "bold",
              }
            }
            testID="box-text"
          >
             
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#4A8305",
              "height": 150,
              "justifyContent": "center",
              "marginBottom": 5,
              "marginLeft": 6,
              "width": 150,
            },
            Object {},
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "opacity": 1,
            }
          }
          testID="box-button"
        >
          <Text
            style={
              Object {
                "fontSize": 40,
                "fontWeight": "bold",
              }
            }
            testID="box-text"
          >
             
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#4A8305",
              "height": 150,
              "justifyContent": "center",
              "marginBottom": 5,
              "marginLeft": 6,
              "width": 150,
            },
            Object {},
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "opacity": 1,
            }
          }
          testID="box-button"
        >
          <Text
            style={
              Object {
                "fontSize": 40,
                "fontWeight": "bold",
              }
            }
            testID="box-text"
          >
             
          </Text>
        </View>
      </View>
    </View>
    <Text
      style={
        Object {
          "color": "#4A8305",
          "fontSize": 30,
          "marginTop": 25,
        }
      }
      testID="player-text"
    >
      Next Player: 
      X
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 0.2,
        "justifyContent": "center",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
              Object {
                "color": "grey",
              },
            ]
          }
        >
          Restart Game
        </Text>
      </View>
    </View>
  </View>
</View>
`;
